src/poliastro/core/perturbations.py

```diff
@@ -204,6 +204,84 @@ def shadow_function(r_sat, r_sun, R):
     return theta > theta_1 + theta_2
 
 
+@jit
+def is_in_umbral_shadow(r_sat, r_sun, r_s, r_p):
+    r"""Calculate whether a satellite is in umbra or not, follows Algorithm 34 from Vallado.
+
+    Parameters
+    ----------
+    r_sat: numpy.ndarray
+        Position of the satellite in the frame of the attractor (km).
+    r_sun: numpy.ndarray
+        Position of the sun in the frame of the attractor (km).
+    r_s: float
+        Radius of the secondary body.
+    r_p: float
+        Radius of the primary body (attractor) responsible for the umbral shadow (km).
+
+    """
+    R_p = norm(r_sun)
+    alpha_um = np.arcsin((r_s - r_p) / R_p)
+    alpha_pen = np.arcsin((r_s + r_p) / R_p)
+
+    dot_sun_sat = np.dot(r_sat, r_sun)
+
+    r_sat_norm = np.sqrt(np.sum(r_sat ** 2))
+    r_sun_norm = np.sqrt(np.sum(r_sun ** 2))
+
+    if dot_sun_sat < 0:
+        angle = np.arccos(dot_sun_sat / r_sat_norm / r_sun_norm)
+        sat_horiz = np.abs(r_sat_norm * np.cos(angle))
+        sat_vert = np.abs(r_sat_norm * np.sin(angle))
+        x = r_p / np.sin(alpha_pen)
+        pen_vert = np.tan(alpha_pen) * (x + sat_horiz)
+        if sat_vert <= pen_vert:
+            y = r_p / np.sin(alpha_um)
+            umb_vert = np.tan(alpha_um) * (y - sat_horiz)
+            # Edge condition for entering umbra: sat_vert - pen_vert = 0.
+            return sat_vert - umb_vert  # +ve to -ve direction means entering umbra.
```

Comment:
What I was thinking is `sat_vert - umb_vert` is the difference in the vertical distances and the sign of that expression could tell whether a satellite is going in or out of a particular shadow. Since the `__call__` method looks for zero crossings, I had thought the return value of these two core functions does give information about the event.

The motivation for doing the above was this line from scipy's `solve_ivp` docs:

> Each event occurs at the zeros of a continuous function of time and state

and also the `LithobrakeEvent` which returns `H - R` inside the `__call__` method.

I was wondering, if we return either 0 or 1, would we lose the ability to tell whether the satellite is going in or out of shadow?

Reply:
I see your point. In the [scipy solve_ivp docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html) it is said that a `direction` attribute can be used to determine the sense in which the zero is found.

However, I think this `direction` is used to determine the limiting cases: the point entering at eclipse and the point going outside of it. The `direction` attribute is in fact another trigger, similar to `terminal`, but focuses on limiting points.

This mean that you are right. By returning only a `0` or a `1` we are loosing the ability to indicate the current `direction` value... Let me think more about how could we face this problem, I will need some time to develop a solution 🤔